parquet/encoding.py

```diff
@@ -1,203 +1,216 @@
-    tup = struct.unpack("<qi", fo.read(12))
-    return tup[0] << 32 | tup[1]
-
-
-def read_plain_float(fo):
-    """Reads a 32-bit float using the plain encoding"""
-    tup = struct.unpack("<f", fo.read(4))
-    return tup[0]
-
+from .thrift_structures import parquet_thrift
 
-def read_plain_double(fo):
-    """Reads a 64-bit float (double) using the plain encoding"""
-    tup = struct.unpack("<d", fo.read(8))
-    return tup[0]
+logger = logging.getLogger("parquet")  # pylint: disable=invalid-name
 
+PY3 = sys.version_info.major > 2
 
-def read_plain_byte_array(fo):
-    """Reads a byte array using the plain encoding"""
-    length = read_plain_int32(fo)
-    return fo.read(length)
+ARRAY_BYTE_STR = u'B' if PY3 else b'B'
 
 
-def read_plain_byte_array_fixed(fo, fixed_length):
-    """Reads a byte array of the given fixed_length"""
-    return fo.read(fixed_length)
+def read_plain_boolean(raw_bytes, count):
+    """Read `count` booleans using the plain encoding."""
+    return np.unpackbits(np.fromstring(raw_bytes, dtype=np.uint8)).reshape(
+            (-1, 8))[:, ::-1].ravel().astype(bool)[:count]
 
-DECODE_PLAIN = {
-    Type.BOOLEAN: read_plain_boolean,
-    Type.INT32: read_plain_int32,
-    Type.INT64: read_plain_int64,
-    Type.INT96: read_plain_int96,
-    Type.FLOAT: read_plain_float,
-    Type.DOUBLE: read_plain_double,
-    Type.BYTE_ARRAY: read_plain_byte_array,
-    Type.FIXED_LEN_BYTE_ARRAY: read_plain_byte_array_fixed
+DECODE_TYPEMAP = {
+    parquet_thrift.Type.INT32: np.int32,
+    parquet_thrift.Type.INT64: np.int64,
+    parquet_thrift.Type.INT96: np.dtype('S12'),
+    parquet_thrift.Type.FLOAT: np.float32,
+    parquet_thrift.Type.DOUBLE: np.float64,
 }
 
 
-def read_plain(fo, type_, type_length):
-    conv = DECODE_PLAIN[type_]
-    if type_ == Type.FIXED_LEN_BYTE_ARRAY:
-        return conv(fo, type_length)
-    return conv(fo)
-
-
-def read_unsigned_var_int(fo):
+def read_plain(raw_bytes, type_, count, width=0):
+    if type_ in DECODE_TYPEMAP:
+        dtype = DECODE_TYPEMAP[type_]
+        return np.frombuffer(memoryview(raw_bytes), dtype=dtype, count=count)
+    if type_ == parquet_thrift.Type.FIXED_LEN_BYTE_ARRAY:
+        dtype = np.dtype('S%i' % width)
+        return np.frombuffer(memoryview(raw_bytes), dtype=dtype, count=count)
+    if type_ == parquet_thrift.Type.BOOLEAN:
+        return read_plain_boolean(raw_bytes, count)
+    # variable byte arrays (rare)
+    file_obj = io.BytesIO(raw_bytes)
+    return np.array([file_obj.read(struct.unpack('<i', file_obj.read(4))[0])
+                     for _ in range(count)], dtype="O")
+
+
+@numba.jit(nogil=True)
+def read_unsigned_var_int(file_obj):
+    """Read a value using the unsigned, variable int encoding.
+    file-obj is a NumpyIO of bytes; avoids struct to allow numba-jit
+    """
     result = 0
     shift = 0
     while True:
-        byte = struct.unpack("<B", fo.read(1))[0]
+        byte = file_obj.read_byte()
```

Comment:
Are there performance implications for reading one byte at a time?  Are we getting around this somehow with NumpyIO buffering and Numba inlining?

Generally it would be good to eventually start benchmarking this implementation against first principles like "how fast are we compared to disk IO limits."